src/optimproblem.cpp

```diff
@@ -737,15 +859,20 @@ void OptimProblem::evalGradF(const Vec x, Vec G){
   double mycost_im = obj_cost_im;
   double myfidelity_re = fidelity_re;
   double myfidelity_im = fidelity_im;
+  double myconstraint = constraint;
+  double my_interm_disc = interm_discontinuity;
   double my_frob2 = frob2;
-  MPI_Allreduce(&mypen, &obj_penal, 1, MPI_DOUBLE, MPI_SUM, comm_init);
+  MPI_Allreduce(&mypen, &obj_penal, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
   MPI_Allreduce(&mypen_dpdm, &obj_penal_dpdm, 1, MPI_DOUBLE, MPI_SUM, comm_init);
-  MPI_Allreduce(&mypenen, &obj_penal_energy, 1, MPI_DOUBLE, MPI_SUM, comm_init);
-  MPI_Allreduce(&mycost_re, &obj_cost_re, 1, MPI_DOUBLE, MPI_SUM, comm_init);
-  MPI_Allreduce(&mycost_im, &obj_cost_im, 1, MPI_DOUBLE, MPI_SUM, comm_init);
-  MPI_Allreduce(&myfidelity_re, &fidelity_re, 1, MPI_DOUBLE, MPI_SUM, comm_init);
-  MPI_Allreduce(&myfidelity_im, &fidelity_im, 1, MPI_DOUBLE, MPI_SUM, comm_init);
-  MPI_Allreduce(&my_frob2, &frob2, 1, MPI_DOUBLE, MPI_SUM, comm_init);
+  MPI_Allreduce(&mypenen, &obj_penal_energy, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+  // Should be comm_init and also comm_time! 
+  MPI_Allreduce(&mycost_re, &obj_cost_re, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
```

Comment:
It is correct because all `mpirank_world` set those variables as zero in the beginning, and only the last time window calculates the actual value.

Ideally, it is good to sum over the last time window of all initial conditions. Currently there is no mpi communicator set up for them. time communicators are split under each initial condition communicator.

Reply:
Yes I think those variables need to be summed up over all initial conditions, and can be summed up over all time window processors since other than the last one set (and leave) it at zero. It needs to be broadcast to ALL processors. Since currently there is no other communicator split (no parallel Petsc), one call to reduce over COMM_WORLD is enough and correct.